zebra-chain/src/value_balance.rs

```diff
@@ -0,0 +1,248 @@
+        // This rule checks the transparent value balance minus the sum of the sprout,
+        // sapling, and orchard value balances in a transaction is nonnegative.
+        (self.transparent - (self.sprout + self.sapling + self.orchard)?)?
+            .constrain::<NonNegative>()
+    }
+
+    /// Creates a `ValueBalance` from the given transparent amount.
+    pub fn from_transparent_amount(transparent_amount: Amount<C>) -> Self {
+        ValueBalance {
+            transparent: transparent_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a `ValueBalance` from the given sprout amount.
+    pub fn from_sprout_amount(sprout_amount: Amount<C>) -> Self {
+        ValueBalance {
+            sprout: sprout_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a `ValueBalance` from the given sapling amount.
+    pub fn from_sapling_amount(sapling_amount: Amount<C>) -> Self {
+        ValueBalance {
+            sapling: sapling_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a `ValueBalance` from the given orchard amount.
+    pub fn from_orchard_amount(orchard_amount: Amount<C>) -> Self {
+        ValueBalance {
+            orchard: orchard_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Insert a transparent value balance into a given `ValueBalance`
+    /// leaving the other values untouched.
+    pub fn transparent_value_balance(
+        &mut self,
+        transparent_value_balance: ValueBalance<C>,
+    ) -> &Self {
+        self.transparent = transparent_value_balance.transparent;
+        self
+    }
+
+    /// Insert a sprout value balance into a given `ValueBalance`
+    /// leaving the other values untouched.
+    pub fn sprout_value_balance(&mut self, sprout_value_balance: ValueBalance<C>) -> &Self {
+        self.sprout = sprout_value_balance.sprout;
+        self
+    }
+
+    /// Insert a sapling value balance into a given `ValueBalance`
+    /// leaving the other values untouched.
+    pub fn sapling_value_balance(&mut self, sapling_value_balance: ValueBalance<C>) -> &Self {
+        self.sapling = sapling_value_balance.sapling;
+        self
+    }
+
+    /// Insert an orchard value balance into a given `ValueBalance`
+    /// leaving the other values untouched.
+    pub fn orchard_value_balance(&mut self, orchard_value_balance: ValueBalance<C>) -> &Self {
+        self.orchard = orchard_value_balance.orchard;
+        self
+    }
+}
+
+impl<C> Default for ValueBalance<C>
+where
+    C: Constraint + Copy,
+{
+    fn default() -> Self {
+        let zero = Amount::zero();
+        Self {
+            transparent: zero,
+            sprout: zero,
+            sapling: zero,
+            orchard: zero,
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug, Clone, PartialEq)]
+//#[allow(missing_docs)]
+/// Errors that can be returned when validating `ValueBalance`s
```

Comment:
Suggestion: just tweaking the comment a bit

```suggestion
/// Errors that can be returned when validating a [`ValueBalance`].
```